File: app/src/main/java/com/chooloo/www/callmanager/CallRecorder.java
package com.chooloo.www.callmanager;


import android.annotation.SuppressLint;
import android.content.Context;
import android.media.MediaRecorder;
import android.os.Build;
import android.os.Environment;
import android.os.Handler;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.chooloo.www.callmanager.util.Utilities;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Objects;

import timber.log.Timber;

public class CallRecorder {

    private static final String LOG_TAG = "CallRecorder";

    // audio constants
    private static final int AUDIO_SOURCE = MediaRecorder.AudioSource.DEFAULT;
    private static final int AUDIO_FORMAT = MediaRecorder.OutputFormat.THREE_GPP;
    private static final int AUDIO_ENCODER = MediaRecorder.AudioEncoder.AMR_NB;
    private static final String FILE_TYPE = "3gpp";

    private String mFileName;
    private boolean mIsRecording = false;
    private Context mContext;
    private MediaRecorder mRecorder;
    private File mFile;

    @RequiresApi(api = Build.VERSION_CODES.O)
    public CallRecorder(Context context, @Nullable String name) {
        mContext = context;

        if (name != null) mFileName = name + '.' + FILE_TYPE;
        else mFileName = generateFileName();

        File downloadPath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
        mFile = new File(downloadPath, mFileName);
        Timber.d("Path of file %s", mFile.getPath());

        try {
            if (mFile.exists()) mFile.delete();
        } catch (Exception e) {
            e.printStackTrace();
        }

        // define recorder
        mRecorder = new MediaRecorder();
        mRecorder.setAudioSource(AUDIO_SOURCE);
        mRecorder.setOutputFormat(AUDIO_FORMAT);
        mRecorder.setAudioEncoder(AUDIO_ENCODER);
        mRecorder.setOutputFile(mFile);
    }

    /**
     * Start recording
     */
    public void start() {
        try {
            mRecorder.prepare();
            mRecorder.start();
            mIsRecording = true;
            Toast.makeText(mContext, "Recording call...", Toast.LENGTH_SHORT).show();
        } catch (IllegalStateException | IOException e) {
            e.printStackTrace();
            Toast.makeText(mContext, "Couldn't start call recording :(", Toast.LENGTH_SHORT).show();
        }

    }

    /**
     * Stop and release recording
     */
    public void stop() {
        if (mRecorder != null) {
            try {
                mRecorder.stop();
                mRecorder.reset();
                mRecorder.release();
                Toast.makeText(mContext, "Call recording ended", Toast.LENGTH_SHORT).show();
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
            mRecorder = null;
        }

        if (mIsRecording) mIsRecording = false;
    }

    @SuppressLint("SimpleDateFormat")
    private static String generateFileName() {
        return "koler_record_" + new SimpleDateFormat("dd-MM-yyyy hh-mm-ss").format(new Date()) + "." + FILE_TYPE;
    }
}

File: app/src/main/java/com/chooloo/www/koler/util/callrecord/CallRecorder.kt
package com.chooloo.www.koler.util.callrecord

import android.annotation.SuppressLint
import android.content.Context
import android.media.MediaRecorder
import android.media.MediaRecorder.AudioSource
import android.widget.Toast
import com.chooloo.www.koler.util.preferences.KolerPreferences
import com.chooloo.www.koler.util.preferences.KolerPreferences.Companion.RecordFormat
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

@SuppressLint("SimpleDateFormat")
class CallRecorder(val context: Context) {
    companion object {
        val DATE_FORMAT = SimpleDateFormat("yy-MM-dd_HH-mm-ss")
        const val OUTPUT_DIR = "Koler Recordings"
    }

    private val _preferences by lazy { KolerPreferences(context) }
    private var _recorder: MediaRecorder? = null
    private var _currentRecording: CallRecording? = null

    val formatChoice: RecordFormat
        get() = _preferences.recordFormat

    fun startRecording(number: String): Boolean {
        if (_recorder != null) {
            stopRecording()
        }
        initRecorder(number)
        return try {
            _recorder?.start()
            true
        } catch (e: Exception) {
            _recorder = null
            false
        }
    }

    fun stopRecording(): CallRecording? {
        _recorder?.apply {
            stop()
            release()
            _recorder = null
        }
        return _currentRecording.also {
            _currentRecording = null
        }
    }

    private fun initRecorder(number: String) {
        val filename = generateFilename(number)
        _recorder = MediaRecorder().apply {
            setAudioSource(AudioSource.VOICE_COMMUNICATION);
            setOutputFormat(formatChoice.outputFormat)
            setAudioEncoder(formatChoice.audioEncoder)
            setOutputFile(filename)
            prepare()
            _currentRecording = CallRecording(
                number = number,
                creationTime = System.currentTimeMillis(),
                filename = filename
            )
        }
    }

    private fun generateFilename(number: String): String {
        val folder = File(context.externalCacheDir, OUTPUT_DIR)
        if (!folder.exists()) {
            folder.mkdirs()
        }
        return ("${folder.absolutePath}${File.separator}Koler_${number}_${DATE_FORMAT.format(Date())}.${formatChoice.encoding}").also {
            Toast.makeText(context, "Saving recording at $it", Toast.LENGTH_LONG).show()
        }
    }
}